File: backend/app/modules/project_activities/schemas.py
"""Project activity pydantic schemas."""
import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

ActivityStatus = Literal["open", "in_progress", "closed"]


class ProjectActivityOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    activity_type_id: uuid.UUID | None
    activity_type_name: str | None
    title: str
    status: str
    assigned_to_id: uuid.UUID | None
    assigned_to_name: str | None
    target_date: date | None
    closed_date: date | None
    remarks: str | None
    sort_order: int
    created_by: uuid.UUID | None
    created_at: datetime
    updated_at: datetime


class ProjectActivityCreate(BaseModel):
    activity_type_id: uuid.UUID | None = None
    title: str = Field(min_length=1, max_length=500)
    status: ActivityStatus = "open"
    assigned_to_id: uuid.UUID | None = None
    target_date: date | None = None
    closed_date: date | None = None
    remarks: str | None = None
    sort_order: int = 0


class ProjectActivityUpdate(BaseModel):
    activity_type_id: uuid.UUID | None = None
    title: str | None = Field(default=None, min_length=1, max_length=500)
    status: ActivityStatus | None = None
    assigned_to_id: uuid.UUID | None = None
    target_date: date | None = None
    closed_date: date | None = None
    remarks: str | None = None
    sort_order: int | None = None

File: frontend/src/features/project-activities/schemas.ts
import { z } from "zod";

export const activityFormSchema = z.object({
  activity_type_id: z.string().uuid().nullable().optional(),
  title: z.string().trim().min(1, "Title is required").max(500),
  status: z.enum(["open", "in_progress", "closed"]).default("open"),
  assigned_to_id: z.string().uuid().nullable().optional(),
  target_date: z.string().optional(),
  remarks: z.string().trim().optional(),
  sort_order: z.coerce.number().int().default(0),
});

export type ActivityFormValues = z.infer<typeof activityFormSchema>;

export const statusOnlySchema = z.object({
  status: z.enum(["open", "in_progress", "closed"]),
  remarks: z.string().trim().optional(),
});

export type StatusOnlyValues = z.infer<typeof statusOnlySchema>;
